orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu

```diff
@@ -77,15 +77,28 @@ __global__ void softmax_warp_backward(output_t* gradInput, const input_t* grad,
   }
 
   acc_t sum[WARP_BATCH];
-#pragma unroll
-  for (int i = 0; i < WARP_BATCH; ++i) {
-    sum[i] = grad_output_reg[i][0];
-#pragma unroll
-    for (int it = 1; it < WARP_ITERATIONS; ++it) {
-      sum[i] += grad_output_reg[i][it];
+  if (!is_log_softmax) {
```

Comment:
I am uncertain about this change... 

See 
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/PersistentSoftmax.cuh#L156

Reply:
For pytorch's backward path, there is only one space where it need to switch backed on is_log_softmax

Reply:
So the is_log_softmax = True expressions are identical between the files. Note the definition of `sum[i]` differs in their case. This expression also matches the one derived earlier for the CPU kernel: `dX_i = d(log Y_i) - [sum_j d(log Y_j)] exp(log Y_i)` for element _i_. It's also not too hard to derive `dX_i = Y_i [ dY_i - sum_j Y_j dY_j  ]` as we use for SoftmaxGrad. Their expression differs by appearance for SoftmaxGrad case. I will think on it.

Reply:
I'm not making sense of the SoftmaxGrad expression from PersistentSoftmax.cuh to be honest.
Their expression is `dX_i = dY_i - Y_i sum_j dY_j`, right?

Consider the specific case `loss := Y_0 + Y_1`. 
Then `Y_0(X_0, X_1) + Y_1(X_0, X_1) = 1` and `dY_0 = dY_1 = 1` and `dX_0 = dX_1 = 0`, right?

That jives with `dX_i = Y_i [ dY_i - sum_j Y_j dY_j ]`: `dX_i = Y_i ( 1 - Y_0 - Y_1 ) = 0.` Check.
It does NOT with `dX_i = dY_i - Y_i sum_j dY_j:` `0 = 1 - 2 Y_i` implies `Y_i = 1/2.` False.

And our expression derives as (.. letting `Z := sum_i exp(X_i)` be the normalization factor..):
```
dX_i := d(loss)/dX_i 
      = sum_j d(loss)/dY_j dY_j/dX_i 
      = sum_j d(loss)/dY_j [ exp(X_i)/Z kronecker_delta_ij - exp(X_j)/Z^2 exp(X_i)]
      = sum_j d(loss)/dY_j [ Y_i kronecker_delta_ij - Y_i Y_j ]
      = d(loss)/dY_i Y_i  - sum_j d(loss)/dY_j Y_i Y_j
      = Y_i [ d(loss)/dY_i - sum_j d(loss)/dY_j Y_j ]
      = Y_i [ dY_i - sum_j dY_j Y_j ]
```

So I'm not sure what's going on in their case.

Reply:
Also see apex [softmax_warp_backward](https://github.com/NVIDIA/apex/blob/4ef930c1c884fdca5f472ab2ce7cb9b505d26c1a/apex/contrib/csrc/multihead_attn/softmax.h#L1606)..